### ISSUE TITLE
Formulate expectations of users and app developers

### ISSUE THREAD

**Issue by rinigus** [poetaster/keypeer.org#6]
I think that other issues (#4 and #1) require agreement regarding expectations of users and app developers. I propose the following:

Keypeer user's expectations:
1. Ability to pay for all services through single donation
2. Pay for used services only
3. Ability to check the balance

App developer expectations:
1. Ability to request any service API keys using user's private token
2. Receiving API keys together with expiration time or "invalid token" response

On keypeer side, it maybe challenging to fulfill these expectations. At least immediately. But, I guess, that's what the funding is applied for. 

Now, these expectations could be also some longer term. However, I am not sure whether subscription model will be easier as we would have to estimate the expenses per month.

**Comment by poetaster** [comment]
Ok, I see this differently. I would avoid bundling. And I would keep unique user tokens for each app. 

In the long term, bundling could 'lift' projects that don't have enough users donating with excess from other projects, but that's a long shot.

**Comment by poetaster** [comment]
The accounting logic that's required if all api keys are dependent on a single user key complicates matters in a way I'd avoid. If we simply have a token per user and application and the knowledge that a payment is made we only have to specify a minimum contribution (for time period X).

I think users should be able to expect:

1.  zero knowledge of their person ends up with api providers
2. almost zero knowledge with keypeer 
3. the payments they make go to the apps they use and support

Application developers should be able to expect:

1. money to pay for an api token suitable for something like commercial (multi-party) use
2. a transparent accounting of which funds came in.

Given the idea of a per person app id/token, application developers would immediately have the accounting information they need so that point 2 is trivial.

**Comment by rinigus** [comment]
Bit difficult to discuss it in two separate issues, but let's do our best :) 

Maybe we should separate long-term goals, or ideal solution, from compromises that are needed to make it take off. So, please specify whether your expectations are the ideal case or a compromise. I assume, it is a compromise and you are fine with the expectations that I formulated for a long term.

I don't see how token per user and application would be better than just token per user. This runs in your response and web page text, so maybe I am missing something. Let's take an example, Pure Maps. In Pure Maps, users would like to get several API keys: Mapbox, HERE, MapTiler, Stadia, and maybe some more. Let's assume that keypeer would support all these providers. How would that work with bundling app and user keys to access keypeer?

In addition, let's say I want to use Mapbox in some desktop map application. It could be Pure Maps or Gnome Maps. Would I have to pay again to use the same service if app and user token are bundled?

Let's go through some of expectations formulated by you:

> 1. zero knowledge of their person ends up with api providers

In principle, map provider would know which tiles some IP address downloaded, searches, and routing. All this is done directly between app and service provider - just API key was fetched from keypeer. So, user would have to trust service provider to handle this information properly. 

> 2. almost zero knowledge with keypeer

Not sure whether it matters - Keypeer should be "simple" and "fair" payment receiver and API keys provider. As it is probably little in it, it doesn't really matter whether you have "zero" or "full" knowledge of it :)

User's expectation 3: agreed. Although, you again use "app" and not "service". For me it is different, not sure whether you make distinction.

Re app developers expectations:

As an app developer, I wouldn't care much of funds transparency. It is an "enabler" for me to get simple access to APIs that users would like. Transparency is probably something that user's would expect.

Re commercial use: that's an interesting question. As providers have different packages, it will probably be service-by-service selection.

Finally:

 > If we simply have a token per user and application and the knowledge that a payment is made we only have to specify a minimum contribution (for time period X).

That will be tricky, I think. We can always start with the bigger fee and then just prolong its use while overall funding lasts.

**Comment by poetaster** [comment]
> Maybe we should separate long-term goals, or ideal solution, from compromises that are needed to make it take off. So, please specify whether your expectations are the ideal case or a compromise. I assume, it is a compromise and you are fine with the expectations that I formulated for a long term.

Yes, I'm trying to formulate a basic set of initial, achievable goals that can be built upon. Keep the complexity low at the beginning and measure the response of the community to see what is required, what needs to be improved, etc.

> I don't see how token per user and application would be better than just token per user. This runs in your response and web page text, so maybe I am missing something.

In #4 you suggest something similar :) It's a question of keeping it simple to match the basic 'flag' paid with one of a number of possible services without a lot of logic. Whether that's realistic or not, I'm not sure. I'm just trying for radicall KISS.

> In addition, let's say I want to use Mapbox in some desktop map application. It could be Pure Maps or Gnome Maps. Would I have to pay again to use the same service if app and user token are bundled?

As I mention, also in #4, this is one of those issues that my scheme does NOT really work for. BUT, I don't think it's relevant for the use case except on the 'edges'. Still, it's an important point in the long term .... So I need to formulate that in the roadmap?



> > 1. zero knowledge of their person ends up with api providers
> 
> In principle, map provider would know which tiles some IP address downloaded, searches, and routing. All this is done directly between app and service provider - just API key was fetched from keypeer. So, user would have to trust service provider to handle this information properly.

Agreed.

> > 3\. almost zero knowledge with keypeer
> 
> Not sure whether it matters - Keypeer should be "simple" and "fair" payment receiver and API keys provider. As it is probably little in it, it doesn't really matter whether you have "zero" or "full" knowledge of it :)

I'm actually trying to keep the whole payments part out of the equation :) That's tricky, but it's why I'm trying to limit the data to a single token per application. Keypeer only needs to know, 'yes this instance is known' and nothing more. The concept I have makes keypeer more an intermediary between an end user and a payment provider than any api provider. The advantage of this approach is that you can swap out payment processing. But it's not even clear if what I imagine is possible at all :)

> Re app developers expectations:
> 
> As an app developer, I wouldn't care much of funds transparency. It is an "enabler" for me to get simple access to APIs that users would like. Transparency is probably something that user's would expect.

Agreed. Number one aim.


> > If we simply have a token per user and application and the knowledge that a payment is made we only have to specify a minimum contribution (for time period X).
> 
> That will be tricky, I think. We can always start with the bigger fee and then just prolong its use while overall funding lasts.

I'm not sure. I'm imagining an ongoing effort that isn't 'just' the payments in the application context but an ongoing overarching project funding. That might be too optimistic or ambitious, but it's in the back of my head.

**Comment by rinigus** [comment]
I will try to keep reply short and address only points that I haven't addressed in #4.

Re long and short term aims. As there is an application for support, there should be ambition as well. It is OK to formulate long term aim and later make the first achievable aims. I suspect that the long-term solution is not that far away and can be done within supported project.

Re use of the same API key by multiple apps: don't think it is such an edge case. Gnome Maps is struggling with the same issues as Pure Maps (both are available at pmOS, Mobian and some other mobile linuxes). Probably the same is with uNav (Ubuntu Touch, Pure Maps is there as well) and OSM Scout (SFOS). So, users would like to be able to just have a single API key that they can use in multiple mapping apps on their platforms.

Re keeping payments out: Not sure how could that be if you are checking some token at opencollective. In #4, token and API keys are addressed in my reply.

Re ongoing project funding: Could be possible. But then it would require some mechanism that would hook to what is provided by keypeer.

**Comment by poetaster** [comment]
> Re use of the same API key by multiple apps: don't think it is such an edge case. Gnome Maps is struggling with the same issues as Pure Maps (both are available at pmOS, Mobian and some other mobile linuxes). Probably the same is with uNav (Ubuntu Touch, Pure Maps is there as well) and OSM Scout (SFOS). So, users would like to be able to just have a single API key that they can use in multiple mapping apps on their platforms.

Ok. Noted. 

I think we need to have a discussion with a group of people that are interested and could play a role. That is, the sailmates. 

One way or another, this is going to need a legal entity to back it. I'll PM you :)

**Comment by b100dian** [comment]
(Sorry if this got long and maybe deviates the discusion, @poetaster feel free to create a new issue from this comment if that's the case)

Would a **mixed** model, like the user just buying its ability to use keypeer for a fixed T period of time work? So:
- user does not donate for specific services, but to access to a keypeer key which
- gives him access to any "onboarded" services for a fixed T **period**, from _any_ mapping app. Ideally the applications store the received original service keys in keyring-like mechanisms.
  (we could ask the user at donation time what services he plans to use, to use this info in the next period(s) - like a poll)
- if donations keep pouring in the next T+1 period could automatically extend existing key usage (donations "spill-out"). If not, new donations are required (as specified initially).
- service allowing finer grained key enrolling (e.g. each user gets his own key) can be used to their full ability, while different key "pooling" strategies would be used for the others.

The period "T" can be defined by looking at the policies for the onboarded services so that keys can be expired after T time and other ones enrolled. For exercise purpose we can imagine something like 3 months for 1EUR.

This would not prevent a rogue user to change the app source code and (ab)use original service keys, but those would expire after T time, and exceeding usage could be charted to reflect higher than anticipated costs -> no donations spill-out can happen.

**Comment by rinigus** [comment]
Trick will be to figure out how much to charge per month, per day or so. But it does sound as a compromise and strategy for moving towards the longer term aims.
